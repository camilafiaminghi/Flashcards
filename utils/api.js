import { AsyncStorage } from 'react-native'
export const FLASHCARDS_STORAGE_KEY = 'Flashcards:decks'

const obj = {
	'JavaScript' : {
		title: 'JavaScript',
		questions: []
	},
	'React' : {
		title: 'React',
		questions: []
	}
}

export function setInitialData (data=obj) {
	// return AsyncStorage.setItem(FLASHCARDS_STORAGE_KEY, JSON.stringify(data))
	return AsyncStorage.mergeItem(FLASHCARDS_STORAGE_KEY, JSON.stringify(data))
}

export function getDecks () {
	return AsyncStorage.getItem(FLASHCARDS_STORAGE_KEY)
}

export function getDeck (entryId) {
	return AsyncStorage.getItem(FLASHCARDS_STORAGE_KEY)
		.then((results) => {
			const data = JSON.parse(results)
			return data[entryId]
		})
}

export function saveDeckTitle ({ title }) {
	return AsyncStorage.mergeItem(FLASHCARDS_STORAGE_KEY, JSON.stringify({
		[title]: {
			title,
			questions: []
		}
	}))
}

export function addCardToDeck (entryId, card) {
	return getDeck(entryId)
		.then((entry) => {
			return AsyncStorage.mergeItem(FLASHCARDS_STORAGE_KEY, JSON.stringify({
				[entryId]: {
					...entry,
					questions: [...entry.questions, card]
				}
			}))
		})
}
