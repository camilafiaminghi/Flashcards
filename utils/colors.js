// utils/colors.js
export const textColor = '#333333'
export const textColorInverse = '#ffffff'
export const pColor = '#9fa8da'
export const pColorLight = '#d1d9ff'
export const pColorDark = '#6f79a8'
export const sColor = '#bdbdbd'
export const sColorLight = '#efefef'
export const sColorDark = '#8d8d8d'

