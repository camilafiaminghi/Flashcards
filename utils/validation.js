export function validationRules (name, value) {
	const rule = {
		title: !!value.trim() && value.length > 6,
		body: !!value.trim() && value.length > 6
	}

	return rule[name]
}

export function isValid (form) {
	return Object.keys(form).every(function (key) {
		return form[key]
	});
}
