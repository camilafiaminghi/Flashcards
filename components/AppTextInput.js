import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { StyleSheet, View, Text, TextInput } from 'react-native'
import { validationRules } from '../utils/validation'
import { textColorInverse, textColor, pColor, pColorLight, pColorDark, sColor, sColorLight, sColorDark } from '../utils/colors'

class AppTextInput extends Component {

	static propTypes = {
		name: PropTypes.string.isRequired,
		placeholder: PropTypes.string.isRequired,
		maxLen: PropTypes.number.isRequired,
		onInputChange: PropTypes.func.isRequired
	}

	state = {
		text: '',
		changed: false,
		valid: false
	}

	handleValidation = (value) => {
		const { name, onInputChange, maxLen } = this.props
		const valid = validationRules(name, value)

		if ( value.length <= maxLen ) {
			this.setState((state) => ({
				...state,
				text: value,
				changed: true,
				valid
			}))

			onInputChange(name, value, valid)
		}
	}

	render() {
		const { text, changed, valid } = this.state
		const { maxLen, placeholder } = this.props
		const textLeft = maxLen - text.length

		return (
			<View style={styles.container}>
				<TextInput
					style={[styles.input, styles.inputText]}
					placeholderTextColor={pColorDark}
	        placeholder={placeholder}
	        onChangeText={this.handleValidation}
	        value={text} />

	      { (textLeft !== 0 && textLeft < maxLen && changed) && <Text style={styles.text}>{ textLeft } characteres left</Text> }
	      { (textLeft === 0 && changed) && <Text style={styles.text}>max characters exceeded</Text>}
			</View>
		)
	}
}

export default AppTextInput

/* STYLES */
const styles = StyleSheet.create({
	container: {
		alignSelf: 'stretch',
		margin: 10
	},
	input: {
		flexDirection: 'row',
		alignSelf: 'stretch',
		height: 50,
		borderBottomColor: pColorDark,
		borderBottomWidth: 2,
		borderRadius: 3
	},
	inputText: {
		padding: 8
	},
	text: {
		fontSize: 12,
		color: textColor
	}
})
