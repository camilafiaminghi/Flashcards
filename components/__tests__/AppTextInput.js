import React, { Component } from 'react'
import { Text, TextInput } from 'react-native'
import renderer from 'react-test-renderer'
import AppTextInput from '../AppTextInput'
import { pColorDark } from '../../utils/colors'

let wrapper
let instance
let component
const props = {
	name: 'title',
	placeholder: 'test',
	maxLen: 10,
	onInputChange: jest.fn()
}

describe('<AppTextInput />', () => {

	beforeEach(() => {
		wrapper = renderer.create(<AppTextInput {...props} />)
		component = shallow(<AppTextInput {...props} />)
	})

	it('matches snapshot', () => {
    expect(wrapper.toJSON()).toMatchSnapshot()
  });

	it('should render', () => {
		expect(wrapper).toBeTruthy()
	})

	it('should show exceeded text onChangeText', () => {
		const input = component.find(TextInput)
		input.props().onChangeText('test value that exceeded the maxlen')

		expect(component.state('changed')).toBe(false)
		expect(component.state('valid')).toBe(false)
	})

	it('should handleValidation onChangeText', () => {
		const input = component.find(TextInput)
		input.props().onChangeText('test value')
		component.update()

		expect(component.state('changed')).toBe(true)
		expect(component.state('valid')).toBe(true)
		expect(props.onInputChange).toHaveBeenCalledTimes(1)
	})

	it('should shows characteres exceeded', () => {
		const input = component.find(TextInput)
		input.props().onChangeText('test value')
		component.update()

		const text = component.find(Text)
		expect(text.shallow().props().children).toEqual('max characters exceeded')
	})

	it('should shows characteres limit', () => {
		const input = component.find(TextInput)
		input.props().onChangeText('test')
		component.update()

		const text = component.find(Text)
		expect(text.shallow().props().children).toEqual([6, ' characteres left'])
	})
})
