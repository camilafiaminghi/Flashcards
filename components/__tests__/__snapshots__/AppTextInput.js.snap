// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppTextInput /> matches snapshot 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "margin": 10,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="test"
    placeholderTextColor="#6f79a8"
    style={
      Array [
        Object {
          "alignSelf": "stretch",
          "borderBottomColor": "#6f79a8",
          "borderBottomWidth": 2,
          "borderRadius": 3,
          "flexDirection": "row",
          "height": 50,
        },
        Object {
          "padding": 8,
        },
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
